cmd/contour/serve_test.go

```diff
@@ -0,0 +1,70 @@
+// Copyright Project Contour Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"testing"
+
+	"github.com/projectcontour/contour/internal/dag"
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/assert"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+func TestGetDAGBuilder(t *testing.T) {
+	t.Run("all default options", func(t *testing.T) {
+		got := getDAGBuilder(newServeContext(), nil, nil, nil, logrus.StandardLogger())
+		assert.Len(t, got.Processors, 4)
+		assert.Empty(t, got.Source.ConfiguredSecretRefs)
+		assert.Equal(t, types.NamespacedName{}, got.Source.ConfiguredGateway)
+		assert.IsType(t, &dag.ListenerProcessor{}, got.Processors[len(got.Processors)-1])
+	})
+
+	t.Run("client cert specified", func(t *testing.T) {
+		clientCert := &types.NamespacedName{Namespace: "client-ns", Name: "client-name"}
+
+		got := getDAGBuilder(newServeContext(), nil, clientCert, nil, logrus.StandardLogger())
+		assert.Len(t, got.Processors, 4)
+		assert.ElementsMatch(t, got.Source.ConfiguredSecretRefs, []*types.NamespacedName{clientCert})
+		assert.Equal(t, types.NamespacedName{}, got.Source.ConfiguredGateway)
+		assert.IsType(t, &dag.ListenerProcessor{}, got.Processors[len(got.Processors)-1])
+	})
+
+	t.Run("fallback cert specified", func(t *testing.T) {
+		fallbackCert := &types.NamespacedName{Namespace: "fallback-ns", Name: "fallback-name"}
+
+		got := getDAGBuilder(newServeContext(), nil, nil, fallbackCert, logrus.StandardLogger())
+		assert.Len(t, got.Processors, 4)
+		assert.ElementsMatch(t, got.Source.ConfiguredSecretRefs, []*types.NamespacedName{fallbackCert})
+		assert.Equal(t, types.NamespacedName{}, got.Source.ConfiguredGateway)
+		assert.IsType(t, &dag.ListenerProcessor{}, got.Processors[len(got.Processors)-1])
+	})
+
+	t.Run("client and fallback certs specified", func(t *testing.T) {
+		clientCert := &types.NamespacedName{Namespace: "client-ns", Name: "client-name"}
+		fallbackCert := &types.NamespacedName{Namespace: "fallback-ns", Name: "fallback-name"}
+
+		got := getDAGBuilder(newServeContext(), nil, clientCert, fallbackCert, logrus.StandardLogger())
+
+		assert.Len(t, got.Processors, 4)
+		assert.ElementsMatch(t, got.Source.ConfiguredSecretRefs, []*types.NamespacedName{clientCert, fallbackCert})
+		assert.Equal(t, types.NamespacedName{}, got.Source.ConfiguredGateway)
+		assert.IsType(t, &dag.ListenerProcessor{}, got.Processors[len(got.Processors)-1])
+	})
+
+	// TODO request/response headers
```

Comment:
Are these tests still in progress or just futures @skriss ?

Reply:
I didn't want to go too nuts before getting some 👀  on the change; let me see if I can cover a few more of these now, and I'll add an issue link in the TODO for anything I'm not gonna cover.

Reply:
awesome, sounds good, just wanted to double check. =)

Reply:
Heh I actually just found a bug in upstream Contour by adding some more test cases here, PR incoming

Reply:
OK, added a couple more cases and added an issue link for the TODO.

Reply:
@stevesloka LMK if you had any other comments, otherwise this should be good to merge.